test/gtest/graphapi_engineheur.cpp

```diff
@@ -0,0 +1,118 @@
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/graphapi/engineheur.hpp>
+
+#include <gtest/gtest.h>
+
+#include "graphapi_gtest_common.hpp"
+
+namespace {
+
+using miopen::graphapi::EngineHeurBuilder;
+using miopen::graphapi::OpGraph;
+
+} // namespace
+
+TEST(GraphApi, EngineHeurBuilder)
+{
+    OpGraph opGraph;
+
+    EXPECT_NO_THROW({
+        EngineHeurBuilder().setOpGraph(&opGraph).setMode(MIOPEN_HEUR_MODE_B).setSmCount(1).build();
+    }) << "EngineHeurBuilder failed on valid attributes";
+
+    EXPECT_NO_THROW({
+        EngineHeurBuilder().setOpGraph(&opGraph).setMode(MIOPEN_HEUR_MODE_B).build();
+    }) << "EngineHeurBuilder failed on valid attributes";
+
+    EXPECT_ANY_THROW({ EngineHeurBuilder().setMode(MIOPEN_HEUR_MODE_B).setSmCount(1).build(); })
+        << "EngineHeurBuilder failed on missing setOpGraph() call";
+
+    EXPECT_ANY_THROW({ EngineHeurBuilder().setOpGraph(&opGraph).setSmCount(1).build(); })
+        << "EngineHeurBuilder failed on missing setMode() call";
+}
+
+namespace {
+
+class MockOpGraphDescriptor : public miopen::graphapi::BackendOperationGraphDescriptor
+{
+public:
+    MockOpGraphDescriptor() { mFinalized = true; }
+};
+
+using miopen::graphapi::GTestDescriptorAttribute;
+using miopen::graphapi::GTestDescriptorSingleValueAttribute;
+using miopen::graphapi::GTestGraphApiExecute;
+
+} // namespace
+
+TEST(GraphApi, EngineHeur)
+{
+    MockOpGraphDescriptor opGraphDescriptor;
+
+    GTestDescriptorSingleValueAttribute<miopenBackendDescriptor_t, char> attrOpGraph(
+        true,
+        "MIOPEN_ATTR_ENGINEHEUR_OPERATION_GRAPH",
+        MIOPEN_ATTR_ENGINEHEUR_OPERATION_GRAPH,
+        MIOPEN_TYPE_BACKEND_DESCRIPTOR,
+        MIOPEN_TYPE_CHAR,
+        2,
+        &opGraphDescriptor);
+
+    GTestDescriptorSingleValueAttribute<miopenBackendHeurMode_t, char> attrMode(
+        true,
+        "MIOPEN_ATTR_ENGINEHEUR_MODE",
+        MIOPEN_ATTR_ENGINEHEUR_MODE,
+        MIOPEN_TYPE_HEUR_MODE,
+        MIOPEN_TYPE_CHAR,
+        2,
+        MIOPEN_HEUR_MODE_B);
+
+    GTestDescriptorSingleValueAttribute<int32_t, char> attrSmCount(
+        true,
+        "MIOPEN_ATTR_ENGINEHEUR_SM_COUNT_TARGET",
+        MIOPEN_ATTR_ENGINEHEUR_SM_COUNT_TARGET,
+        MIOPEN_TYPE_INT32,
+        MIOPEN_TYPE_CHAR,
+        2,
+        1);
+
+    GTestGraphApiExecute<GTestDescriptorAttribute*> execute;
+
+    execute.descriptor.attrsValid = false;
+    execute.descriptor.textName   = "MIOPEN_BACKEND_ENGINEHEUR_DESCRIPTOR";
+    execute.descriptor.type       = MIOPEN_BACKEND_ENGINEHEUR_DESCRIPTOR;
+    execute.descriptor.attributes = {&attrOpGraph, &attrSmCount};
+    execute();
+
+    execute.descriptor.attributes = {&attrMode, &attrSmCount};
+    execute();
+
+    execute.descriptor.attrsValid = true;
+    execute.descriptor.attributes = {&attrOpGraph, &attrMode, &attrSmCount};
+    execute();
+}
```

Comment:
I noticed that the code does not prevent the user from _changing_ attributes. This is being tested implicitly by reusing the `execute` object, but we should switch to _explicit_, tests for all cases.

I started playing with it and found that the test support does not behave the same way the library does. See `OpGraphModeAModeBBroken` and `ModeBOpGraphModeABroken` below.

I also split the tests up; otherwise, when one does fail you aren't given any indication which individual test fails. The gtest output references the test support code.

```
class EngineHeur : public testing::Test
{
protected:
    MockOpGraphDescriptor                                                   opGraphDescriptor;
    GTestDescriptorSingleValueAttribute<miopenBackendDescriptor_t, char>    attrOpGraph;
    GTestDescriptorSingleValueAttribute<miopenBackendHeurMode_t, char>      attrModeA;
    GTestDescriptorSingleValueAttribute<miopenBackendHeurMode_t, char>      attrModeB;
    GTestDescriptorSingleValueAttribute<int32_t, char>                      attrSmCount;
    GTestGraphApiExecute<GTestDescriptorAttribute*>                         execute;

    void SetUp() override
    {
        attrOpGraph = GTestDescriptorSingleValueAttribute<miopenBackendDescriptor_t, char> {
            true,
            "MIOPEN_ATTR_ENGINEHEUR_OPERATION_GRAPH",
            MIOPEN_ATTR_ENGINEHEUR_OPERATION_GRAPH,
            MIOPEN_TYPE_BACKEND_DESCRIPTOR,
            MIOPEN_TYPE_CHAR,
            2,
            &opGraphDescriptor
        };

        attrModeA = GTestDescriptorSingleValueAttribute<miopenBackendHeurMode_t, char> {
            true,
            "MIOPEN_ATTR_ENGINEHEUR_MODE",
            MIOPEN_ATTR_ENGINEHEUR_MODE,
            MIOPEN_TYPE_HEUR_MODE,
            MIOPEN_TYPE_CHAR,
            2,
            MIOPEN_HEUR_MODE_A
        };

        attrModeB = GTestDescriptorSingleValueAttribute<miopenBackendHeurMode_t, char> {
            true,
            "MIOPEN_ATTR_ENGINEHEUR_MODE",
            MIOPEN_ATTR_ENGINEHEUR_MODE,
            MIOPEN_TYPE_HEUR_MODE,
            MIOPEN_TYPE_CHAR,
            2,
            MIOPEN_HEUR_MODE_B
        };

        attrSmCount = GTestDescriptorSingleValueAttribute<int32_t, char> {
            true,
            "MIOPEN_ATTR_ENGINEHEUR_SM_COUNT_TARGET",
            MIOPEN_ATTR_ENGINEHEUR_SM_COUNT_TARGET,
            MIOPEN_TYPE_INT32,
            MIOPEN_TYPE_CHAR,
            2,
            1
        };

        execute.descriptor.textName   = "MIOPEN_BACKEND_ENGINEHEUR_DESCRIPTOR";
        execute.descriptor.type       = MIOPEN_BACKEND_ENGINEHEUR_DESCRIPTOR;
    }
};

// TODO: you could save some lines by using TEST_P; however, I personally prefer the descriptive test names

TEST_F(EngineHeur, NoModeFails)
{
    execute.descriptor.attrsValid = false;
    execute.descriptor.attributes = {&attrOpGraph, &attrSmCount};
    execute();
}
TEST_F(EngineHeur, NoOpGraphFails)
{
    execute.descriptor.attrsValid = false;
    execute.descriptor.attributes = {&attrModeA, &attrSmCount};
    execute();
}
TEST_F(EngineHeur, OpGraphModeAValid)
{
    execute.descriptor.attrsValid = true;
    execute.descriptor.attributes = {&attrOpGraph, &attrModeA, &attrSmCount};
    execute();
}
TEST_F(EngineHeur, ModeBOpGraphValid)
{
    execute.descriptor.attrsValid = true;
    execute.descriptor.attributes = {&attrModeB, &attrOpGraph, &attrSmCount};
    execute();
}
TEST_F(EngineHeur, OpGraphOpGraphFails)
{
    execute.descriptor.attrsValid = false;  // "replace" OpGraph: false is correct
    execute.descriptor.attributes = {&attrOpGraph, &attrOpGraph, &attrSmCount};
    execute();
}
TEST_F(EngineHeur, OpGraphOpGraphModeValid)
{
    execute.descriptor.attrsValid = true;   // "replace" OpGraph plus Mode: true is correct
    execute.descriptor.attributes = {&attrOpGraph, &attrOpGraph, &attrModeA, &attrSmCount};
    execute();
}
TEST_F(EngineHeur, OpGraphModeAModeBBroken) // TODO: change test name when fixed
{
    execute.descriptor.attrsValid = true;   // change Mode: incorrect, always fails whether true or false
    execute.descriptor.attributes = {&attrOpGraph, &attrModeA, &attrModeB, &attrSmCount};
    execute();
}
TEST_F(EngineHeur, ModeBOpGraphModeABroken) // TODO: change test name when fixed
{
    execute.descriptor.attrsValid = false;   // change Mode: incorrect, always fails whether true or false
    execute.descriptor.attributes = {&attrModeB, &attrOpGraph, &attrModeA, &attrSmCount};
    execute();
}
```

Reply:
It wasn't intended to prevent a user to change attributes. A user provides a vector of attributes as a sequence of attributes he goes to set and test the behavior.

We were using TEST_P combined with custom print function (for provided sequence of attributes) to run _explicit_ cases. But yes it seems to be not that easy to understand the output. To switch to individual TEST it might be convenient to add a more suitable class than `execute`

Reply:
Having separate tests is a good idea IMO.